### PR TITLE
Refactor checkModify flags into an enum

### PR DESCRIPTION
**Pull request by RazvanN7**
cc @ibuclaw

### CODE REVIEW

File: src/dmd/expression.d [hunk 1]
```diff
@@ -94,6 +94,15 @@ enum Modifiable
     initialization,
 }
 
+// Specifies how the checkModify deals with certain situations
```
Comment (dkorpel) [inline src/dmd/expression.d:97]: This can be a DDoc comment

**Review by dkorpel — APPROVED** [review]

File: src/dmd/expression.d [hunk 2]
```diff
@@ -4818,7 +4827,7 @@ extern (C++) final class DotVarExp : UnaExp
                                 /* checkModify will consider that this is an initialization
                                  * of v while it is actually an assignment of a field of v
                                  */
-                                scope modifyLevel = v.checkModify(loc, sc, dve.e1, !onlyUnion ? (flag | 2) : flag);
+                                scope modifyLevel = v.checkModify(loc, sc, dve.e1, !onlyUnion ? (flag | ModifyFlags.fieldAssign) : flag);
```
Comment (ibuclaw) [inline src/dmd/expression.d:4830]: Should the callers of `checkModifiable` be updated as well to pass `ModifyFlags.noError`?  In which case, the `flag` variable could also be switched over to the enum type.

Comment (RazvanN7) [reply src/dmd/expression.d:4833]: Done. I'm surprised how little the flag is used. I had to modify a single call site; the rest just go with the default.

**Review by ibuclaw — DISMISSED** [review]
Headers need to be in sync first, then this is fine to be merged.

**Comment by RazvanN7** [comment]
@ibuclaw Done.

File: src/dmd/expression.h [hunk 3]
```diff
@@ -95,7 +108,7 @@ class Expression : public ASTNode
     virtual bool checkType();
     virtual bool checkValue();
     bool checkDeprecated(Scope *sc, Dsymbol *s);
-    virtual int checkModifiable(Scope *sc, int flag = 0);
+    virtual int checkModifiable(Scope *sc, ModifyFlags flag =ModifyFlags.none);
```
Comment (ibuclaw) [inline src/dmd/expression.h:111]:
```suggestion
    virtual int checkModifiable(Scope *sc, ModifyFlags flag = ModifyFlags::none);
```

Comment (ibuclaw) [reply src/dmd/expression.h]: These functions could really do with being devirtualized and made `extern(D)` in continuation of #12809 and others.

Comment (RazvanN7) [reply src/dmd/expression.h]: Will look into it.

**Comment by RazvanN7** [comment]
@ibuclaw can you please merge so I can devirtualize these functions?

**Pull request merged**